### PR TITLE
build with go 1.17

### LINKED ISSUE #24533: Upgrade Golang to v1.17
#### Feature request description

Upgrade Golang to version 1.17 
https://golang.org/doc/go1.17

#### Is your feature request related to a problem? If so, please describe.

No

#### Describe alternatives you've considered.

NA

#### Additional context

NA

### PR DESCRIPTION
**Pull request by camdencheek**
Figured I'd get the ball rolling on this since there seem to be some potentially exciting perf gains in 1.17. I've been using 1.17 locally since beta, and haven't run into any issues. 

Besides, when @efritz [gives you advice](https://sourcegraph.slack.com/archives/C3B3SDBMY/p1630082244002500?thread_ts=1630081967.001100&cid=C3B3SDBMY), you should probably take it. 

Closes #24533

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -206,6 +206,120 @@ require (
 	sigs.k8s.io/yaml v1.2.0 // indirect
 )
 
+require (
```
Comment (efritz) [inline go.mod:209]: does `go mod tidy` change this new list?

Comment (camdencheek) [reply go.mod:209]: Yes -- the only thing I changed manually was line 3. This was all `go mod tidy`.

Comment (camdencheek) [reply go.mod:209]: It is kinda weird, isn't it?

Comment (bobheadxi) [reply go.mod:209]: I guess go1.17 sorts indirect dependencies into a new `require` block. kind of nice actually

Comment (bobheadxi) [reply go.mod:209]: it doesnt look like they're removed from the other `require` block though so not sure what's up with that 🤔

Comment (efritz) [reply go.mod:209]: Appreciated if true!

Comment (bobheadxi) [reply go.mod:209]: https://golang.org/doc/go1.17#go-command

> If a module specifies go 1.17 or higher in its go.mod file, its go.mod file now contains an explicit require directive for every module that provides a transitively-imported package. (In previous versions, the go.mod file typically only included explicit requirements for directly-imported packages.)

This doesn't look like what's happening here, huh

Comment (camdencheek) [reply go.mod:209]: https://github.com/golang/go/issues/47733

Comment (bobheadxi) [reply go.mod:209]: https://github.com/golang/go/issues/47733#issuecomment-907705692 this behaviour looks like fun 👀

Comment (camdencheek) [reply go.mod:209]: Seems like it's safe to remove them manually from the first require block?

**Review by eseliger — APPROVED** [review]

**Review by unknwon — APPROVED** [review]

**Review by mrnugget — APPROVED** [review]

**Review by ryanslade — APPROVED** [review]

**Comment by camdencheek** [comment]
Any concerns from @sourcegraph/distribution?

**Review by bobheadxi — APPROVED** [review]
LGMT

**Pull request merged**